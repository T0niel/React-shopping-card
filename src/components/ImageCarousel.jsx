import { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

export default function ImageCarousel({ imgs }) {
  const [index, setIndex] = useState(0);
  const [loading, setLoading] = useState(true); 

  useEffect(() => {
    imgs.forEach((img) => {
      const image = new Image();
      image.src = img.src; // Image starts loading and gets cached
      img.decoding = 'async';
    });
  }, [imgs]);

  useEffect(() => {
    setLoading(true);
  }, [index]);

  const handleImageLoad = () => {
    setLoading(false);
  };

  const handlePrevClick = () => {
    setIndex((prevIndex) => (prevIndex - 1 + imgs.length) % imgs.length);
  };

  const handleNextClick = () => {
    setIndex((prevIndex) => (prevIndex + 1) % imgs.length);
  };

  if (imgs.length === 0) return <div>No images available</div>;

  return (
    <div className="w-[100%] h-[100%] relative border-2 rounded border-gray-200">
      {loading && (
        <div className="loader absolute top-0 left-0 right-0 bottom-0 -z-10 flex justify-center items-center m-[auto]"></div>
      )}

      {index < imgs.length && (
        <img
          src={imgs[index].src}
          alt={imgs[index].alt}
          className={`object-contain w-[100%] h-[100%] ${
            loading ? 'opacity-50' : 'opacity-100'
          }`}
          onLoad={handleImageLoad}
        />
      )}
      {imgs.length > 1 && (
        <>
          <button
            className="absolute top-0 h-[100%] w-8 font-bold text-xl"
            onClick={handlePrevClick}
          >
            <div className="w-[100%] h-[100%] bg-gray-400 rounded hover:opacity-40 transition-opacity ease-in delay-100 opacity-20 absolute top-0"></div>
            {'<'}
          </button>
          <button
            className="absolute top-0 right-0 h-[100%] w-8 font-bold text-xl"
            onClick={handleNextClick}
          >
            <div className="w-[100%] h-[100%] bg-gray-400 rounded hover:opacity-40 transition-opacity ease-in delay-100 opacity-20 absolute top-0"></div>
            {'>'}
          </button>
        </>
      )}
    </div>
  );
}




ImageCarousel.propTypes = {
  imgs: PropTypes.arrayOf(
    PropTypes.shape({
      src: PropTypes.string,
      alt: PropTypes.string,
    })
  ),
};
