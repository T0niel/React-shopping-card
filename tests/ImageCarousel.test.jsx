import { describe, expect, it } from 'vitest';
import ImageCarousel from '../src/components/ImageCarousel';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('ImageCarousel', () => {
  it('Matches snapshot', () => {
    const {container} = render(<ImageCarousel imgs={[]}/>);

    expect(container).toMatchSnapshot()
  })

  it('Displays the first image', () => {
    const imgs = [
      { src: 'https://example.com/imgA', alt: 'imageA' },
      { src: 'https://example.com/imgB', alt: 'imageB' },
      { src: 'https://example.com/imgB', alt: 'imageC' },
    ];
    render(<ImageCarousel imgs={imgs} />);

    const img = screen.getByRole('img');
    expect(img.src).toEqual('https://example.com/imgA');
  })

  it('Adds alt to the first image', () => {
    const imgs = [
      { src: 'https://example.com/imgA', alt: 'imageA' },
      { src: 'https://example.com/imgB', alt: 'imageB' },
      { src: 'https://example.com/imgB', alt: 'imageC' },
    ];
    render(<ImageCarousel imgs={imgs} />);

    const img = screen.getByRole('img');
    expect(img.alt).toEqual('imageA');
  });

  it('Has a button with the text ">" that spins to the images on the right when clicked', async () => {
    const user = userEvent.setup();

    const imgs = [
      { src: 'https://example.com/imgA', alt: 'imageA' },
      { src: 'https://example.com/imgB', alt: 'imageB' },
      { src: 'https://example.com/imgC', alt: 'imageC' },
    ];
    render(<ImageCarousel imgs={imgs} />);

    const img = screen.getByRole('img');
    const btn = screen.getByText('>');

    expect(img.src).toEqual('https://example.com/imgA');
    expect(img.alt).toEqual('imageA');
    await user.click(btn);
    expect(img.src).toEqual('https://example.com/imgB');
    expect(img.alt).toEqual('imageB');
    await user.click(btn);
    expect(img.src).toEqual('https://example.com/imgC');
    expect(img.alt).toEqual('imageC');

    //When clicked again it wraps around
    await user.click(btn);
    expect(img.src).toEqual('https://example.com/imgA');
    expect(img.alt).toEqual('imageA');
  });

  it('Has a button with the text "<" that spins to the images on the left when clicked', async () => {
    const user = userEvent.setup();

    const imgs = [
      { src: 'https://example.com/imgA', alt: 'imageA' },
      { src: 'https://example.com/imgB', alt: 'imageB' },
      { src: 'https://example.com/imgC', alt: 'imageC' },
    ];
    render(<ImageCarousel imgs={imgs} />);

    const img = screen.getByRole('img');
    const btn = screen.getByText('<');

    //When clicked again it wraps around
    await user.click(btn);
    expect(img.src).toEqual('https://example.com/imgC');
    expect(img.alt).toEqual('imageC');
    await user.click(btn);
    expect(img.src).toEqual('https://example.com/imgB');
    expect(img.alt).toEqual('imageB');
    await user.click(btn);
    expect(img.src).toEqual('https://example.com/imgA');
    expect(img.alt).toEqual('imageA');

  });
});
