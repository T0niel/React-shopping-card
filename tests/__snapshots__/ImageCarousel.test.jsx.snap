// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImageCarousel > Matches snapshot 1`] = `
<div>
  <div>
    No images available
  </div>
</div>
`;
